### PR TITLE
Cleanup curve parameter access

### LINKED ISSUE #249: Unused values in recover_pubkey_from_signed_message
**Describe the bug**
Some variables are unused after initialization.

**Environment information:**

- OS: Any
- Platform: Any
- Firmware Version: 46ef0d346a835263d0c060d074652698954613f7

**Steps to Reproduce**

Look at https://github.com/skycoin/hardware-wallet/blob/46ef0d346a835263d0c060d074652698954613f7/skycoin-api/skycoin_check_signature.c#L96

**Actual behavior**
After putting some values in [`r` and `s`](https://github.com/skycoin/hardware-wallet/blob/46ef0d346a835263d0c060d074652698954613f7/skycoin-api/skycoin_check_signature.c#L97-L98) these are not used.

**Expected behavior**
Maybe these values should be used.

PD: Maybe there is not a bug at all but it's recommended to get a review from a crypto-expert

### LINKED ISSUE #255: Include certain missing skycoin cipher lib test vectors
**Feature description**

Describe the feature
The tests are possibly missing some important test vectors for the cipher libs

This is the test case that must be imported:
https://github.com/skycoin/skycoin/blob/develop/src/cipher/secp256k1-go/secp256_test.go#L654-L738

**Is your feature request related to a problem? Please describe.**
Testing/sanity checking

**Describe the solution you'd like**
Add the test vectors to the test suite

**Describe alternatives you've considered**
No alternatives

**Additional context**
I've discovered some historically important test vectors in underlying secp256k1 libraries that may not have been included in the libskycoin test suite because it is in a deeper library than top-level `cipher`.  The initial discussion of them is here https://github.com/piotrnar/gocoin/issues/15

**Possible implementation**
Preliminary ideas to get this done.
If you have all details in mind then provide a checklist

- [ ] Translate the test case 
- [ ] Confirm that they pass. If they don't pass, then there is a problem with the libskycoin cipher library

### LINKED ISSUE #262: test_signature only checks the first 32 bytes of the signature
The tests in `test_signature` only check the first 32 bytes of the signature.  If there is a reason for that, it should be explained why, otherwise the full signature should be checked, and it should match the result of the skycoin core libs.

### LINKED ISSUE #263: ecdsa_skycoin_sign nonce should be 32 bytes, not 32 bits
**Describe the bug**

secp256k1 nonces should be 32 bytes, not 32 bits

**Possible implementation**

Use `bignum256`

### PR DESCRIPTION
**Pull request by gz-c**
Fixes #249 
Fixes #263 
Fixes #262 
Fixes #255 
Related to #104 

Changes:	

- Remove erroneous use of `HDNode` as a roundabout way of obtaining the secp256k1 curve parameters
- Remove unnecessary duplicated code in `skycoin_check_signature_tools.{c,h}`
- Remove unused values in `recover_pubkey_from_signed_message`
- Remove decred's curve (we have no intention of supporting decred)
- Add secp256k1.o into the bootloader bin. This could cause the bootloader size to be too large, please check it
- Revert to using the ecdsa_sign_digest function from the ecdsa library, instead of a butchered copy of that function
- Modify ecdsa_sign_digest to grant control of random k values for testing
- Update the `protob` submodule 
- Add ecdh tests
- Rewrite skycoin signature verification routine, to use the underlying ecdsa library method

 Does this change need to mentioned in CHANGELOG.md?	
no	

 Requires testing	
yes

 Comments about testing , should you have some

- Will need to retest firmware signing and validation, since those routines were modified, and there is no automated test for them
- Added more test vectors from skycoin core, but its possible more could be added

### CODE REVIEW

File: skycoin-api/skycoin_check_signature.c [hunk 1]
```diff
@@ -44,9 +49,10 @@ int verify_digest_recover(const ecdsa_curve* curve, uint8_t* pub_key, const uint
     }
 
     memcpy(&cp.x, &r, sizeof(bignum256));
+
     // compute y from x
-    uncompress_mcoords(curve, recid & 1, &cp.x, &cp.y);
-    if (!mecdsa_validate_pubkey(curve, &cp)) {
+    uncompress_coords(curve, recid & 1, &cp.x, &cp.y);
+    if (!ecdsa_validate_pubkey(curve, &cp)) {
```
Comment (stdevAlDen) [inline skycoin-api/skycoin_check_signature.c]: @gz-c this opens the need to link `skycoin-crypto/tools/ecdsa.c` in the bootloader too?

Comment (olemis) [reply skycoin-api/skycoin_check_signature.c]: looks like so ...

**Comment by stdevAlDen** [comment]
@gz-c [`curve_info`](https://github.com/gz-c/hardware-wallet/blob/2c52c4d887197554b50f8ab583ee35405e9dc83b/skycoin-api/tools/bip32.h#L34-L38) and [`ecdsa_curve`](https://github.com/gz-c/hardware-wallet/blob/2c52c4d887197554b50f8ab583ee35405e9dc83b/skycoin-api/tools/ecdsa.h#L37-L49) looks like different data types, maybe I'm wrong but please check if make sense assigning a [`curve_info`](https://github.com/gz-c/hardware-wallet/blob/2c52c4d887197554b50f8ab583ee35405e9dc83b/skycoin-api/tools/secp256k1.c#L80) to a [`ecsda_curve`](https://github.com/gz-c/hardware-wallet/blob/2c52c4d887197554b50f8ab583ee35405e9dc83b/skycoin-api/skycoin_check_signature.c#L23) if this is a mistake I need to know why the compiler does not fire us because this? any idea @olemis?

**Comment by stdevAlDen** [comment]
> @gz-c [Travis test failures](https://travis-ci.com/skycoin/hardware-wallet/jobs/206702766#L3337-L3347).

@gz-c maybe this can help https://github.com/skycoin/hardware-wallet/pull/254/files#r292133634

File: skycoin-api/skycoin_check_signature.c (hunk 1 above)
Comment (gz-c) [reply skycoin-api/skycoin_check_signature.c]: Added

**Comment by gz-c** [comment]
@olemis why am i getting "build failed" emails but github is showing it succeeded? in travis i see one build passed, the other failed

**Comment by olemis** [comment]
@gz-c will take a look , but make full-firmware was failing due to symbols not found by ld

**Comment by olemis** [comment]
> @olemis why am i getting "build failed" emails but github is showing it succeeded? in travis i see one build passed, the other failed

oh ! I see , weird ...

**Comment by olemis** [comment]
> @olemis why am i getting "build failed" emails but github is showing it succeeded?

well ... i seems [build 883](https://travis-ci.com/skycoin/hardware-wallet/builds/115030830) succeeded .

File: skycoin-api/tools/ecdsa.c [hunk 2]
```diff
@@ -696,93 +697,148 @@ int ecdsa_sign_double(const ecdsa_curve* curve, HasherType hasher_type, const ui
...
+    // message digest to sign must not be 0
+    if (bn_is_zero(&z)) {
+    	return -2;
+    }
+
     for (i = 0; i < 10000; i++) {
-        // generate random number k
+        // generate random number k (nonce)
         generate_k_random(&k, &curve->order);
+	    // generate another random number randk to randomize operations
```
Comment (olemis) [inline skycoin-api/tools/ecdsa.c]: @gz-c some minor whitespace issues in this block of code

**Comment by gz-c** [comment]
@olemis do you know any problem with using the full 65 byte signature in the bootloader, instead of the 64 signature?

**Comment by olemis** [comment]
> @olemis do you know any problem with using the full 65 byte signature in the bootloader, instead of the 64 signature?

I'm not aware of any such issues , but we're testing the patch . If we find something , let you know asap .

File: skycoin-api/tools/ecdsa.c (hunk 2 above)
Comment (gz-c) [reply skycoin-api/tools/ecdsa.c]: Fixed

File: fsm_impl.c [hunk 3]
```diff
@@ -0,0 +1,582 @@
+/*
```
Comment (stdevAlDen) [inline fsm_impl.c]: @gz-c do you copy this file to the root dir due to any particular reason?

**Pull request merged**